src/lib/crypto/kimchi_backend/common/plonk_constraint_system.ml

```diff
@@ -0,0 +1,1244 @@
+    | Snarky_backendless.Constraint.Boolean v -> (
+        let s, x = red v in
+        match x with
+        | `Var x ->
+            (* -x + x * x = 0  *)
+            add_generic_constraint ~l:x ~r:x
+              [| Fp.(negate one); Fp.zero; Fp.zero; Fp.one; Fp.zero |]
+              sys
+        | `Constant ->
+            assert (Fp.(equal s (s * s))) )
+    | Snarky_backendless.Constraint.Equal (v1, v2) -> (
+        let (s1, x1), (s2, x2) = (red v1, red v2) in
+        match (x1, x2) with
+        | `Var x1, `Var x2 ->
+            (* s1 x1 - s2 x2 = 0
+          *)
+            if not (Fp.equal s1 s2) then
+              add_generic_constraint ~l:x1 ~r:x2
+                [| s1; Fp.(negate s2); Fp.zero; Fp.zero; Fp.zero |]
+                sys
+              (* TODO: optimize by not adding generic costraint but rather permuting the vars *)
+            else
+              add_generic_constraint ~l:x1 ~r:x2
+                [| s1; Fp.(negate s2); Fp.zero; Fp.zero; Fp.zero |]
+                sys
+        | `Var x1, `Constant ->
+            add_generic_constraint ~l:x1
+              [| s1; Fp.zero; Fp.zero; Fp.zero; Fp.negate s2 |]
+              sys
+        | `Constant, `Var x2 ->
+            add_generic_constraint ~r:x2
+              [| Fp.zero; s2; Fp.zero; Fp.zero; Fp.negate s1 |]
+              sys
+        | `Constant, `Constant ->
+            assert (Fp.(equal s1 s2)) )
+    | Plonk_constraint.T (Basic { l; r; o; m; c }) ->
+        (* 0
+           = l.s * l.x
+           + r.s * r.x
+           + o.s * o.x
+           + m * (l.x * r.x)
+           + c
+           =
+             l.s * l.s' * l.x'
+           + r.s * r.s' * r.x'
+           + o.s * o.s' * o.x'
+           + m * (l.s' * l.x' * r.s' * r.x')
+           + c
+           =
+             (l.s * l.s') * l.x'
+           + (r.s * r.s') * r.x'
+           + (o.s * o.s') * o.x'
+           + (m * l.s' * r.s') * l.x' r.x'
+           + c
+        *)
+        (* TODO: This is sub-optimal *)
+        let c = ref c in
+        let red_pr (s, x) =
+          match red x with
+          | s', `Constant ->
+              c := Fp.add !c Fp.(s * s') ;
+              (* No need to have a real term. *)
+              (s', None)
+          | s', `Var x ->
+              (s', Some (Fp.(s * s'), x))
+        in
+        (* l.s * l.x
+           + r.s * r.x
+           + o.s * o.x
+           + m * (l.x * r.x)
+           + c
+           =
+             l.s * l.s' * l.x'
+           + r.s * r.x
+           + o.s * o.x
+           + m * (l.x * r.x)
+           + c
+           =
+        *)
+        let l_s', l = red_pr l in
+        let r_s', r = red_pr r in
+        let _, o = red_pr o in
+        let var = Option.map ~f:snd in
+        let coeff = Option.value_map ~default:Fp.zero ~f:fst in
+        let m =
+          match (l, r) with
+          | Some _, Some _ ->
+              Fp.(l_s' * r_s' * m)
+          | _ ->
+              (* TODO: Figure this out later. *)
```

Comment:
I suspect we may have to do this soon, given the incoming updated random oracle input.